DynamisHelper: refactor and update with additional Treasury changes

Skyrant opened the pull request:
Added nicer interface with more options.  works with dynamis campaign
event. general refactoring of code and optimization.

Skyrant committed:
refactor and update

Added nicer interface with more options.  works with dynamis campaign
event. general refactoring of code and optimization.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -27,19 +27,307 @@
+-- Parse the chat messages for drops, staggers and time extensions ------------
+-------------------------------------------------------------------------------
+windower.register_event('incoming text',function (original, new, color)
+	a,b,fiend = string.find(original,"%w+'s attack staggers the (%w+)%!")
```
Consider making these local.

Skyrant replied:
done

Skyrant committed:
updated stagger timer

stagger timer now displays the mobs name and the stun icon.

Skyrant committed:
added time extention tracker

displays which time extensions you still need removing them once
obtained.

Skyrant committed:
compatibility to old version

commands now compatible to the original version.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -1,4 +1,4 @@
---Copyright (c) 2013, Krizz
+--Copyright (c) 2013, Krizz, Skyrant
```
Please change the `(c)` to `©`.

Skyrant replied:
done

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,22 +24,427 @@
+
+-- upgrade to new settings file if old settings still exist -------------------
+if settings.trposx and settings.trposy then
+	-- copy position from old config to new one
```
Throughout the file you're mixing spaces and tabs for indentation. Please use 4 spaces uniformly.

Skyrant replied:
fixed

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,22 +24,427 @@
+-- Parse the chat messages for drops, staggers and time extensions ------------
+-------------------------------------------------------------------------------
+windower.register_event('incoming text',function (original, new, color)
+	a,b,fiend = string.find(original,"%w+'s attack staggers the (%w+)%!")
```
Actually, it seems like you never care about `a` or `b`, only for `fiend`. If that's the case, you should use `string.match` instead:
`local fiend = original:match('%w+'s attack staggers the (%w+)%!')`

You should only use `string.find` if you need the start and end index of the match.

And this goes for all the other `a,b,x` lines in this function as well.

Skyrant replied:
done

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,22 +24,427 @@
+obj_time = 0
+end_time = 0
+
+image = texts.new("image")
```
I don't know if you actually want this, but if you do, you can make text object changes automatically save to the settings file. So every time you change the position for example, it will automatically save.

To do that, you need to call the constructor like this:
```lua
image = texts.new('image', settings, settings)
```

This will make the settings you currently have obsolete, you'd need to change them to the ones the text library uses. But in return you get many features for free, and the user can completely customize the text box if they want to. You can try it once and you'll see what settings it will generate. If you go that route I'd recommend putting the text settings into its own sub-table:
```lua
defaults = {}
defaults.image = {}

-- and later

image = texts.new('image', settings.image, settings)
```

So the first argument is the settings table for the image object, the second argument is for the root settings object.

Skyrant replied:
investigating, is there some documentation for the texts library?

z16 replied:
Unfortunately no, but there's a short explanation of the [constructor](https://github.com/Windower/Lua/blob/dev/addons/libs/texts.lua#L129) in the Lua file itself. Check the link for more info.

Skyrant committed:
variable uninitialized

a variale caused an error because it was not initialized

Skyrant committed:
recommended fixes

fixed whitespace, identation, local variables, string matching

Skyrant committed:
Added new command

Added new command (boolean citydrop) that prevents itesm from being
dropped when the player is in a city (eg. buying from vendor, finishing
quests etc.).

Skyrant committed:
texts lib - fixed 2 errors

stroke_width and stroke_transparency were using the wrong settings.

Skyrant committed:
Revert "texts lib - fixed 2 errors"

This reverts commit 3ecffaa3db64399afe462ecd4cd1bf117416b6de.

Skyrant committed:
Merge remote-tracking branch 'Windower/dev' into dev

Skyrant committed:
DynamisHelper complete rewrite

Uses texts library and has many more functions compared to the original.

Skyrant committed:
Fix - initialize time extensions on zone change

Time extensions were not reset on zone change.

z16 commented on addons/Treasury/Treasury.lua:
```diff
@@ -103,7 +104,15 @@ function force_check()
+                end
+            end
+        end
+    elseif settings.AutoDrop then
```
This entire block you added can be merged with the previous block if you just change the condition:
```lua
    if settings.AutoDrop and (settings.CityDrop or not areas.Cities:contains(res.zones[windower.ffxi.get_info().zone].name)) then
```

z16 commented on addons/Treasury/Treasury.lua:
```diff
@@ -183,8 +192,13 @@ windower.register_event('incoming chunk', function(id, data)
+            else
+                stack()
+            end
+        elseif id == 0x020 and settings.AutoDrop and code.drop:contains(chunk.Item) then
```
Can be merged same as above:
```lua
        if id == 0x020 and settings.AutoDrop and code.drop:contains(chunk.Item) and (settings.CityDrop or not areas.Cities:contains(res.zones[windower.ffxi.get_info().zone].name)) then
```

z16 commented on addons/Treasury/Treasury.lua:
```diff
@@ -276,6 +290,17 @@ windower.register_event('addon command', function(command1, command2, ...)
+
+        config.save(settings)
+        log('CityDrop %s':format(settings.CityDrop and 'enabled' or 'disabled'))
+        --log('Current Zone: %s':format(res.zones[windower.ffxi.get_info().zone].name))
```
I guess that was just for debugging, can be removed.

z16 commented on addons/Treasury/Treasury.lua:
```diff
@@ -324,6 +350,35 @@ windower.register_event('addon command', function(command1, command2, ...)
 end)
 
+areas = {}
+areas.Cities = S{
```
```lua
areas.cities
```

A common style please :)

z16 reviewed (changes requested):
The changes to Treasury are fine, although the checks can be consolidated, but that's optional. But please submit a different pull request from the DynamisHelper one. They need to be reviewed and handled separately, which is difficult if they're in the same request.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,597 @@
+function init_currency()
+    Currency = T{"Ordelle Bronzepiece","Montiont Silverpiece","One Byne Bill",
+                "One Hundred Byne Bill","Tukuku Whiteshell","Lungo-Nango Jadeshell",
+                "Forgotten Thought","Forgotten Hope","Forgotten Touch","Forgotten Journey","Forgotten Step"}
```
Since you're using T-tables you can use their advanced iteration features:
```lua
    for c in Currency:it() do
        w[c] = 0
    end
```

Same goes for the `init_granules` function below.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,597 @@
+-- Did we enter a Dynamis Zone? -----------------------------------------------
+-------------------------------------------------------------------------------
+windower.register_event('zone change', function(zone)
+    if ProcZones:contains(windower.ffxi.get_info().zone) then
```
No need to call `windower.ffxi.get_info` here, just to get the zone; it's already a function argument.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,597 @@
+-------------------------------------------------------------------------------
+windower.register_event('prerender', function()
+    if obj_time < 1 then return end
+    if obj_time ~= (end_time - os.time()) then
```
Can merge these conditions:
```lua
    if obj_time < 1 or obj_time == end_time - os.time() then
        return
    end

    --[[ rest of the function ]]
```

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,597 @@
+    window.appendline(window,"\\cr————————————————————")
+    window.appendline(window,"${current_mob|(unknown)}\n"..Green.."${current_proc|(none)}")
+    window.appendline(window,"\\cr————————————————————")
+    for i=1, #Currency do
```
See above. As you're using the Currency table now it doesn't need to be a T-table at all.

z16 commented on addons/dynamishelper/dynamishelper.lua:
```diff
@@ -24,408 +24,597 @@
+-------------------------------------------------------------------------------
+-- Process options and save settings ------------------------------------------
+-------------------------------------------------------------------------------
+windower.register_event('addon command',function (...)
```
Instead of using `:lower()` dozens of times below, you can use it just once at the top by designing the function with a more fitting signature:
```lua
windower.register_event('addon command', function(command, ...)
    command = command and command:lower() or 'help'
    local options = {...}
```

This will replace the first 7 lines of the function, and you can remove all `:lower()` calls on command below

z16 reviewed (commented):
I added the review for the DynamisHelper changes as well, but like I said before, I'd really appreciate if you could split it up into two pull requests.
